### ISSUE TITLE
Custom Topic Naming Strategy

### ISSUE THREAD

**Issue by embano1** [awslabs/eventbridge-kafka-connector#27]
## What is your idea?

This was suggested by @baldawar to make the current topic to `detail-type` naming more flexible for users (instead of the current hardcoded behavior which provides certain flexibility with the `${topic}` placeholder.

The idea is to use a class e.g., `TopicDetailTypeMapper`, providing a default implementation and make it configurable via `aws.eventbridge.detail.types.mapper.class`.

## Would you be willing to make the change? 

No

## Additional context

This would replace the current `aws.eventbridge.detail.types` configuration, providing the current default value as the default behavior of the new class. Users with a custom configuration would be warned with a `BREAKING CHANGE` in the release notes.

**Comment by embano1** [comment]
Another benefit: if different event types are within the same topic, a custom topic naming strategy (class) would be able to set specific event types (`detail-type`) on the outgoing events to EventBridge, making IAM-based subscription and filtering controls even more powerful with EventBridge for Kafka users.

**Comment by embano1** [comment]
@maschnetwork see my latest comment, this would be HUGE for users!

**Comment by maschnetwork** [comment]
@embano1 Yes should be able to start this in Jan.

**Comment by embano1** [comment]
Great! Let's discuss UX/design here once you have some cycles. 

@agebhar1 this will be the first user-settable custom class implementation, so definitely keen to keep you in the loop here.

**Comment by maschnetwork** [comment]
@embano1 @agebhar1 After diving a bit deeper and looking for the abstraction I realized that we already abstracted the mapping process via [EventBridgeMapper](https://github.com/awslabs/eventbridge-kafka-connector/blob/main/src/main/java/software/amazon/event/kafkaconnector/mapping/EventBridgeMapper.java). If we would make this class configurable one could already implement a custom detail/topic mapping functionality. I agree it would not be as straight forward as only the single method but you also get more flexibility by changing headers for example. Wdyt?

**Comment by embano1** [comment]
IMHO that's quiet some code you have to write just to extract/modify detail-type mapping. But you raise a good point: we also have a public method which does this in the mapper: getDetailType in Config. 

How about making that an interface?

**Comment by maschnetwork** [comment]
Yes that was my initial thought (if we agree that EventBridgeMapper abstraction is too broad). 

Another point, what do you think about keeping the `aws.eventbridge.detail.types` and base the new default implementation on the current implementation while mentioning in the docs you can overwrite this behavior with `aws.eventbridge.detail.types.mapper.class`? This would not be a breaking change and IMO is still transparent. This would give users the current flexibility via the config and if they need something more advanced they implement the class.

**Comment by embano1** [comment]
> Yes that was my initial thought (if we agree that EventBridgeMapper abstraction is too broad).

+1 but let's see what @agebhar1 thinks

> Another point, what do you think about keeping the aws.eventbridge.detail.types and base the new default implementation on the current implementation while mentioning in the docs you can overwrite this behavior with aws.eventbridge.detail.types.mapper.class? This would not be a breaking change and IMO is still transparent. This would give users the current flexibility via the config and if they need something more advanced they implement the class.

IIUC, that was exactly the idea: keep it backwards compatible so that:

- if neither `...class` nor `aws.eventbridge.detail.types` are set: use current default behavior of the existing (default) implementation
- if `aws.eventbridge.detail.types` is set, apply this using the existing (default) implementation
- if `...class` is provided, ignore `aws.eventbridge.detail.types` and the behavior is driven by the provided class

**Comment by embano1** [comment]
One issue could be though that this `class` likely needs to also know about the deserialization logic if it is going to access the Kafka record `key` or `value`

**Comment by maschnetwork** [comment]
> IIUC, that was exactly the idea: keep it backwards compatible so that:

Alright then it makes sense - the issue description read different:

> This would replace the current aws.eventbridge.detail.types configuration, providing the current default value as the default behavior of the new class. Users with a custom configuration would be warned with a BREAKING CHANGE in the release notes.

**Comment by embano1** [comment]
Fair, `replace` is probably not the right word then anymore :)

**Comment by agebhar1** [comment]
To be sure what we are talking about.

https://github.com/awslabs/eventbridge-kafka-connector/blob/e748150ee051eaebba9b4a51192ac01c23f8abf9/src/main/java/software/amazon/event/kafkaconnector/mapping/DefaultEventBridgeMapper.java#L53-L67

Exactly this assignment (line 60) `.detailType(config.getDetailType(record.topic()))` , where 
https://github.com/awslabs/eventbridge-kafka-connector/blob/e748150ee051eaebba9b4a51192ac01c23f8abf9/src/main/java/software/amazon/event/kafkaconnector/EventBridgeSinkConfig.java#L170-L174
is.

**Comment by agebhar1** [comment]
>>  Yes that was my initial thought (if we agree that EventBridgeMapper abstraction is too broad).

> +1 but let's see what @agebhar1 thinks
+1

**Comment by agebhar1** [comment]
+1 for an (functional) interface to map from `SinkRecord` to detail type `String` which can be configured in the `DefaultEventBridgeMapper` and is by default the current implementation

**Comment by embano1** [comment]
> To be sure what we are talking about.

yes, that innocent line `.detailType(config.getDetailType(record.topic()))`

> Maybe in addition to SinkRecord the plain (original) configuration Map<?, ?> could also be provided for the mapping arguments (BiFunction).

Can you please elaborate on the idea/use case behind?

**Comment by agebhar1** [comment]
After further consideration, the assignment with the additional configuration no longer makes sense. So the mapping function signature might be `SinkRecord -> String` (DetailType). 

But instead of having the configuration in the mapping function, there could be a method to provide the connection configuration or a filtered subset to configure the custom class instead of hard coding it. e.g.
```
public void configure(Map<String, ?> configs) 
```
like the [Configurable](https://kafka.apache.org/35/javadoc/org/apache/kafka/common/Configurable.html) interface of Kafka. The [ConsumerInterface](https://kafka.apache.org/35/javadoc/org/apache/kafka/clients/consumer/ConsumerInterceptor.html) for example implements it.

**Comment by embano1** [comment]
> So the mapping function signature might be SinkRecord -> String (DetailType)
Yes

> to provide the connection configuration
Which connection? I'm not 100% following

IIUC that, if a user provides a custom class implementation for `SinkRecord -> String` there is a case where they want to configure this class as well, i.e., also needing signature `public void configure(Map<String, ?> configs) `?

**Comment by agebhar1** [comment]
Sorry connector not connection configuration.

> IIUC that, if a user provides a custom class implementation for SinkRecord -> String there is a case where they want to configure this class as well, i.e., also needing signature public void configure(Map<String, ?> configs) ? Which I agree! So instead of BiFunction, the custom class extends Configurable

Yes, you did.

**Comment by maschnetwork** [comment]
> But instead of having the configuration in the mapping function, there could be a method to provide the connection configuration or a filtered subset to configure the custom class instead of hard coding it. e.g.

+1 I like that approach. Will start working on a draft shortly.

> What I don't know is how they'd pass their own properties to the connector config? Will it just work if they follow the same approach we do in EventBridgeSinkConfig? For example, when I just add "foo.bla":"something" to our connector config JSON today and submit it to Kafka Connect, anything not defined in EventBridgeSinkConfig is ignored/dropped (?)

I don't think that this is needed. Technically one could leverage the existing `aws.eventbridge.detail.types` for any config that the mapper can leverage. Providing custom configuration IMO is beyond the scope of this issue.

**Comment by embano1** [comment]
> I don't think that this is needed.

Fair point, IMHO we can start with a functional interface and always expand if needed. I suspect the usage here will be converting the `SinkRecord` to a concrete structured data type, extract a field to then return a `detail-type` mapping e.g., if the `SinkRecord` contains an event type already or as part of the record's `key`.

**Comment by embano1** [comment]
It would be great if you could capture that scenario as part of your unit tests so we can discuss the ergonomics/UX of the idea.

**Comment by maschnetwork** [comment]
@agebhar1 @maschnetwork Added first Draft PR for further discussion / comments. Please take a look.

**Comment by agebhar1** [comment]
I will hopefully give you feedback in the next few days.

**Comment by agebhar1** [comment]
I did @maschnetwork :+1: 

Question: How is the custom class accessible to the Kafka Connect JVM? Is it a custom build of this connector or a separate Jar?

**Comment by embano1** [comment]
> Question: How is the custom class accessible to the Kafka Connect JVM? Is it a custom build of this connector or a separate Jar?

Do you mean the custom class != our default implementation? Assuming so, yes separate JAR.

**Comment by agebhar1** [comment]
Custom class.

**Comment by maschnetwork** [comment]
Converted the Draft PR to PR. Any additional suggestions?

**Comment by embano1** [comment]
@maschnetwork thx!

There's one ask I had which would be really valuable: now that we have refactored the topic-detail mapper, giving users options to implement their own and providing a default using our current implementation, I strongly believe we should also ship a custom implementation which users can directly use instead of writing their own for a common use case:

_"As a user, I want to specify a field from my `SinkRecord` which will be used in the outgoing `detail-type`"_

Now, this requires a bit of work, I agree, since we have to use SerDes to convert the record into JSON and then use for example `JSONPath` (which @agebhar1 is introducing) to extract a key.

So my proposal would be to ship your PR, in parallel develop a custom implementation using `JSONPath`, and perhaps a bit of refactoring to access the internal JSON conversion logic in our connector to avoid repetition. We would ship this custom implementation in the upcoming release.

Thoughts?

**Comment by maschnetwork** [comment]
@embano1 I like the idea but I would track this is a separate issue and start discussion there. Wdyt?

**Comment by embano1** [comment]
+1

**Comment by maschnetwork** [comment]
Seems like this was not closed yet.
